Add ability to use challenges other than Route53 DNS

felddy opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->
- Add the ability to use other Certbot challenges other than the default Route53.
- Upstream changes to implement `ghcr.io` container repository.

By adding the `--no-dns-route53` as the first parameter to the container, it will disable the default Route53 challenge.
This allows other challenges to be used.  e.g.; `--manual` 

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

We have a use case where we need to get certificates for domains that are not (yet) hosted in Route53.  

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

@dav3r is testing this with an external domain.
- [x] @dav3r says "it's gud".

Tested in CI.

## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

felddy committed:
Enable container publication to GitHub Container Registry.

* Duplicates DockerHub tags for ghcr.io
* Adds login to GitHub Container Registry using the GITHUB_TOKEN secret.
* Documentation update to reflect multi-registry publication.

felddy committed:
Update .github/workflows/build.yml

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

felddy committed:
Merge branch 'improvement/gh_registry' of github.com:cisagov/skeleton-docker into improvement/manual-override

dav3r commented:
Successful test!

```
❱ docker-compose -f my-docker-compose.yml run certboto --no-dns-route53 certonly --manual --preferred-challenges dns -d mysuperfakedomain.gov -d mail.mysuperfakedomain.gov -d www.mysuperfakedomain.gov
Syncing certbot configs from my-certificates-bucket
Rebuilding symlinks in /etc/letsencrypt
Route53 DNS challenge disabled by --no-dns-route53 flag
Running: certbot certonly --manual --preferred-challenges dns -d mysuperfakedomain.gov -d mail.mysuperfakedomain.gov -d www.mysuperfakedomain.gov
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator manual, Installer None
Requesting a certificate for mysuperfakedomain.gov and 2 more domains
Performing the following challenges:
dns-01 challenge for mail.mysuperfakedomain.gov
dns-01 challenge for mysuperfakedomain.gov
dns-01 challenge for www.mysuperfakedomain.gov

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Please deploy a DNS TXT record under the name:

_acme-challenge.mail.mysuperfakedomain.gov.

with the following value:

cmV1AljDcvVJjcv-0hgOAAvWt3CJaHtuS8G3b_ss3zw

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Press Enter to Continue

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Please deploy a DNS TXT record under the name:

_acme-challenge.mysuperfakedomain.gov.

with the following value:

SuDA0J4HDFK_pNEX7CfI5yLJi0LgYL9NkdTY3UDDu8s

(This must be set up in addition to the previous challenges; do not remove,
replace, or undo the previous challenge tasks yet. Note that you might be
asked to create multiple distinct TXT records with the same name. This is
permitted by DNS standards.)

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Press Enter to Continue

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Please deploy a DNS TXT record under the name:

_acme-challenge.www.mysuperfakedomain.gov.

with the following value:

UMVtRnz1mC4Sc_eWHCe6gebyNiuYyug27kDp-0qnO4k

(This must be set up in addition to the previous challenges; do not remove,
replace, or undo the previous challenge tasks yet. Note that you might be
asked to create multiple distinct TXT records with the same name. This is
permitted by DNS standards.)

Before continuing, verify the TXT record has been deployed. Depending on the DNS 
provider, this may take some time, from a few seconds to multiple minutes. You can
check if it has finished deploying with aid of online tools, such as the Google
Admin Toolbox: https://toolbox.googleapps.com/apps/dig/#TXT/_acme-challenge.www.mysuperfakedomain.gov.
Look for one or more bolded line(s) below the line ';ANSWER'. It should show the
value(s) you've just added.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Press Enter to Continue
Waiting for verification...
Cleaning up challenges

IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at:
   /etc/letsencrypt/live/mysuperfakedomain.gov/fullchain.pem
   Your key file has been saved at:
   /etc/letsencrypt/live/mysuperfakedomain.gov/privkey.pem
   Your certificate will expire on 2021-08-11. To obtain a new or
   tweaked version of this certificate in the future, simply run
   certbot again. To non-interactively renew *all* of your
   certificates, run "certbot renew"
 - If you like Certbot, please consider supporting our work by:

   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le

Syncing certbot configs to my-certificates-bucket
upload: ../../etc/letsencrypt/archive/mysuperfakedomain.gov/cert1.pem to s3://my-certificates-bucket/archive/mysuperfakedomain.gov/cert1.pem
upload: ../../etc/letsencrypt/archive/mysuperfakedomain.gov/chain1.pem to s3://my-certificates-bucket/archive/mysuperfakedomain.gov/chain1.pem
upload: ../../etc/letsencrypt/archive/mysuperfakedomain.gov/privkey1.pem to s3://my-certificates-bucket/archive/mysuperfakedomain.gov/privkey1.pem
upload: ../../etc/letsencrypt/archive/mysuperfakedomain.gov/fullchain1.pem to s3://my-certificates-bucket/archive/mysuperfakedomain.gov/fullchain1.pem
upload: ../../etc/letsencrypt/csr/0203_csr-certbot.pem to s3://my-certificates-bucket/csr/0203_csr-certbot.pem
upload: ../../etc/letsencrypt/keys/0203_key-certbot.pem to s3://my-certificates-bucket/keys/0203_key-certbot.pem
upload: ../../etc/letsencrypt/live/mysuperfakedomain.gov/fullchain.pem to s3://my-certificates-bucket/live/mysuperfakedomain.gov/fullchain.pem
upload: ../../etc/letsencrypt/live/mysuperfakedomain.gov/chain.pem to s3://my-certificates-bucket/live/mysuperfakedomain.gov/chain.pem
upload: ../../etc/letsencrypt/live/mysuperfakedomain.gov/privkey.pem to s3://my-certificates-bucket/live/mysuperfakedomain.gov/privkey.pem
upload: ../../etc/letsencrypt/live/mysuperfakedomain.gov/cert.pem to s3://my-certificates-bucket/live/mysuperfakedomain.gov/cert.pem
upload: ../../etc/letsencrypt/renewal/mysuperfakedomain.gov.conf to s3://my-certificates-bucket/renewal/mysuperfakedomain.gov.conf
upload: ../../etc/letsencrypt/live/mysuperfakedomain.gov/README to s3://my-certificates-bucket/live/mysuperfakedomain.gov/README
```

dav3r reviewed (approved):
Manual override- what a useful feature, thanks for adding it!  ⏏️

jsf9k commented on .pre-commit-config.yaml:
```diff
@@ -60,6 +60,9 @@ repos:
     rev: 1.0.5
     hooks:
       - id: shell-lint
+        # TODO: SC2039 is retired in 0.7.2 but not 0.7.0.  Local checks will
+        # pass but they will currently fail in GitHub Actions which uses 0.7.0
+        args: [-x, --exclude=SC2039]
```
Can you create an issue to remove this when possible and include a link to it here?  One of the checkboxes in the standard PR template references doing this, so it is required.

felddy replied:
At @mcdonnnj 's suggestion I've implemented this differently so that it won't require a disable or a TODO.

felddy replied:
Some additional context about `SC2039`: 
https://github.com/koalaman/shellcheck/wiki/SC2039

And an issue about `SC30xx` codes not being properly reported:
https://github.com/koalaman/shellcheck/issues/2052

jsf9k replied:
@felddy - normally we let the reporter resolve comments.

jsf9k reviewed (changes requested):
I requested one minor change, and then I will heartily approve.  Please re-request my approval after addressing it.

felddy committed:
Exclude shellcheck SC2039 and SC3040.

SC2039 is retired in 0.7.2 but not 0.7.0.  It has been migrated to new 
SC30** checks.  Local checks will pass but 
they will currently fail in GitHub Actions which uses 0.7.0

felddy committed:
Add ability to disable Route53 DNS challenge.

felddy commented:
@dav3r : Just saw `mysuperfakedomain`  teehee

jsf9k commented on src/entrypoint.sh:
```diff
@@ -4,7 +4,7 @@ set -o nounset
 # Sha-bang cannot be /bin/bash (not available), but
 # the container's /bin/sh does support pipefail.
-# shellcheck disable=SC2039
+# shellcheck disable=SC2039,SC3040
```
What do you think about adding [these links](https://github.com/cisagov/certboto-docker/pull/22#discussion_r632088000) to this comment?

felddy replied:
Good suggestion.  Added in https://github.com/cisagov/certboto-docker/pull/22/commits/37276f7a78f38a9cff62dba5c9a410b08943161e

jsf9k reviewed (approved):
Approved, but @felddy please see my one suggestion.

mcdonnnj reviewed (approved):
A solid addition. 💪💪💪

felddy committed:
Clarify why both spellcheck codes are listed.

felddy committed:
Bump version.
